### PR TITLE
[SPARK-45452][SQL] Improve `InMemoryFileIndex` to use `FileSystem.listFiles` API

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to use `InMemoryFileIndex` to use `FileSystem.listFiles` API instead of many recursive FileSystem calls.

To do this, a new configuration is introduced with the default value, `s3a`.

```
spark.sql.sources.useListFilesFileSystemList=s3a
```

### Why are the changes needed?

To improve the nested directory listing performance. Currently, it's designed to apply for a single root directory listing case on S3A file system. For example, the improvement on a 3-year-data table with `year/month/day/hour` hierarchy is **158x**.

**3-YEAR PARTITIONED TABLE (2021 ~ 2023)**
```
$ bin/spark-sql
spark-sql (default)> CREATE TABLE t USING CSV PARTITIONED BY (year, month, day, hour) LOCATION "file:///tmp/three_year" AS SELECT /*+ REPARTITION(64, year, month, day, hour) */ Y.id year, M.id month, D.id day, H.id hour, 0 value FROM RANGE(2021, 2024) Y, RANGE(1, 13) M, RANGE(1, 31) D, RANGE(0, 24) H;

$ aws s3 sync three_year s3://dongjoon/three_year

$ aws s3 ls s3://dongjoon/three_year --summarize --recursive | more
2023-10-07 01:07:00          8 three_year/._SUCCESS.crc
2023-10-07 01:07:00          0 three_year/_SUCCESS
2023-10-07 01:07:00         12 three_year/year=2021/month=1/day=1/hour=0/.part-00033-264e1d55-d466-43c8-8972-283a1f84082b.c000.csv.crc
2023-10-07 01:07:00          2 three_year/year=2021/month=1/day=1/hour=0/part-00033-264e1d55-d466-43c8-8972-283a1f84082b.c000.csv
...

$ aws s3 ls s3://dongjoon/three_year --summarize --recursive | tail -n2
Total Objects: 51842
   Total Size: 362888
```

**AFTER (11s)**
```
$ bin/spark-shell \
-c spark.hadoop.fs.s3a.aws.credentials.provider=com.amazonaws.auth.profile.ProfileCredentialsProvider
...
scala> spark.time(new org.apache.spark.sql.execution.datasources.InMemoryFileIndex(spark, Seq(new org.apache.hadoop.fs.Path("s3a://dongjoon/three_year")), Map.empty, None))
Time taken: 11270 ms
```

**BEFORE (1741s)**
```
$ bin/spark-shell \
-c spark.hadoop.fs.s3a.aws.credentials.provider=com.amazonaws.auth.profile.ProfileCredentialsProvider \
-c spark.sql.sources.useListFilesFileSystemList=none
...
scala> spark.time(new org.apache.spark.sql.execution.datasources.InMemoryFileIndex(spark, Seq(new org.apache.hadoop.fs.Path("s3a://dongjoon/three_year")), Map.empty, None))
Time taken: 1741315 ms
```

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the CIs with the newly added test case.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this PR when you get a chance, @viirya and @sunchao ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 1]
```diff
@@ -1641,6 +1641,16 @@ object SQLConf {
       .booleanConf
       .createWithDefault(false)
 
+  val USE_LISTFILES_FILESYSTEM_LIST =
+    buildConf("spark.sql.sources.useListFilesFileSystemList")
+      .doc("A comma-separated list of file system schems to use FileSystem.listFiles API " +
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1646]: schemes?

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala [hunk 2]
```diff
@@ -69,6 +69,38 @@ private[spark] object HadoopFSUtils extends Logging {
       ignoreMissingFiles, ignoreLocality, parallelismThreshold, parallelismMax)
   }
 
+  /**
+   * Lists a collection of paths recursively with a single API invocation.
+   * Like parallelListLeafFiles, this ignores FileNotFoundException on the given root path.
+   *
+   * This is able to be be called on both driver and executors.
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:76]:
```suggestion
   * This is able to be called on both driver and executors.
```

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala [hunk 3]
```diff
@@ -308,4 +340,14 @@ private[spark] object HadoopFSUtils extends Logging {
...
   }
+
+  /** Checks if we should filter out this path. */
+  def shouldFilterOutPath(path: String): Boolean = {
+    val exclude = (path.contains("/_") && !path.contains("=")) || path.contains("/.")
+    val include = path.contains("/_common_metadata/") ||
+      path.endsWith("/_common_metadata") ||
+      path.contains("/_metadata/") ||
+      path.endsWith("/_metadata")
+    (exclude && !include) || shouldFilterOutPathName(path)
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:351]: Isn't `shouldFilterOutPathName` already covering these patterns?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 4]
```diff
@@ -150,16 +150,28 @@ object InMemoryFileIndex extends Logging {
...
-      ignoreMissingFiles =
-        new FileSourceOptions(CaseInsensitiveMap(parameters)).ignoreMissingFiles,
-      ignoreLocality = sparkSession.sessionState.conf.ignoreDataLocality,
-      parallelismThreshold = sparkSession.sessionState.conf.parallelPartitionDiscoveryThreshold,
-      parallelismMax = sparkSession.sessionState.conf.parallelPartitionDiscoveryParallelism)
+    val fileSystemList =
+      sparkSession.sessionState.conf.useListFilesFileSystemList.split(",").map(_.trim)
+    val ignoreMissingFiles =
+      new FileSourceOptions(CaseInsensitiveMap(parameters)).ignoreMissingFiles
+    val useListFiles = paths.size == 1 &&
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala:157]: For multiple input paths, can we also use listFiles on each of them and merge returned list?

**Review by viirya — COMMENTED** [review]
That looks great improvement. I have a few minor comments so far.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:1646]: Thanks!

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:76]: Thank you. I'll update with the above typo together.

File: core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala (hunk 3 above)
Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/util/HadoopFSUtils.scala:351]: - `shouldFilterOutPathName` only handles `file name`. 
- `shouldFilterOutPath` handles the intermediate directory in the path like `s3://abc/_metadata/1.txt`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 4 above)
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala:157]: Yes, we will add that layer. Currently, this PR is designed to optimize for a single table operation like `CREATE TABLE` and `REPARE TABLE`.

**Comment by dongjoon-hyun** [comment]
Thank you, @viirya !

**Comment by dongjoon-hyun** [comment]
The last commit is about fixing two typos in the string and comment. And, the previous commit passed all tests. Let me merge this for Apache Spark 4.0.0. Thank you again, @viirya .

Screenshot 2023-10-07 at 4 51 17 PM

**Comment by sunchao** [comment]
late LGTM, the improvement looks great!

**Comment by dongjoon-hyun** [comment]
Thank you, @sunchao .

**Comment by cloud-fan** [comment]
Have we tested it on a flattened directory layout (no partitions but many files)? Does it have perf regression?

**Comment by cloud-fan** [comment]
Just FYI, we hit a 10x perf regression with this change in our internal benchmark with a flattened directory layout. This PR says it can replace many recursive FileSystem calls, which is not an issue for flattened directories. Shall we disable it by default? Or we only use it when scanning a file source table and the table metadata says it has more than one partition column.

**Comment by dongjoon-hyun** [comment]
Thank you for sharing and sorry for being late, @cloud-fan .  I was on a vacation from December 14th.

Could you elaborate a little more specific? For example, how many files do you have in a single S3 prefix? What is the exact time value *before* and *after* when you say `10x`?
> Just FYI, we hit a 10x perf regression with this change in our internal benchmark with a flattened directory layout. 

Of course, we can, @cloud-fan . Before doing that, I want to reproduce your benchmark to make it sure and to improve for that case.
> Shall we disable it by default?

**Comment by cloud-fan** [comment]
The time went from 10 seconds to more than 1 minute. I don't know the exact number of files but it should be a lot as it took 10 seconds to list files.

**Comment by dongjoon-hyun** [comment]
As you see in this PR description, the listing time of 51842 files was 11 seconds in my experiment.
> For example, the improvement on a 3-year-data table with year/month/day/hour hierarchy is 158x.

Given your description, I guess `100s` (the 10x result) means about `520k` files. In other words, did you try to list 0.5M files or 1M files in a single S3 prefix? Could you double-check with the internal benchmark team, please?

In addition, in your benchmark case, do you have some other S3 page size or S3-compatible layer like DBFS in your benchmark environment? It could have a side-effect due to the layered file system.

Please shed some lights to us. 😄 I'd like to provide a proper action on this topic because this is really helpful for the most of  tables and users.

**Comment by cloud-fan** [comment]
> As you see in this PR description, the listing time of 51842 files was 11 seconds in my experiment.

Did you try with a flattened directory?

**Comment by dongjoon-hyun** [comment]
Not yet. I'm trying to understand your case.

BTW, may I ask what do you mean by `a flattened directory`? So far, I thought what you mean is a table on a directory (without any nesting or partitions) with 1M files? Is it some other technique?

**Comment by cloud-fan** [comment]
> is a table on a directory (without any nesting or partitions) with 1M files?

yea, no partition directories.

**Comment by dongjoon-hyun** [comment]
To @cloud-fan , I tested like the following in 100k scale. I don't see any **10x** regression which you mentioned. Given this result, could you provide your specific procedure? Otherwise, I don't think Apache Spark has an issue.

1. Prepare **100k files** in a single directory
```bash
$ mkdir /tmp/100k

$ for i in {1..100000}; do touch /tmp/100k/$i.txt; done

$ aws s3 sync /tmp/100k s3://dongjoon/100k

$ aws s3 ls s3://dongjoon/100k/ --summarize | tail -n2
Total Objects: 100000
   Total Size: 0
```

2. Build Apache Spark 4
```
$ NO_MANUAL=1 ./dev/make-distribution.sh -Phadoop-cloud,hive
```

3. Comparison

**Apache Spark 3.5.0**
```
$ bin/spark-shell \
-c spark.hadoop.fs.s3a.aws.credentials.provider=com.amazonaws.auth.profile.ProfileCredentialsProvider
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
24/01/04 04:02:36 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
24/01/04 04:02:37 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
Spark context Web UI available at http://localhost:4041
Spark context available as 'sc' (master = local[*], app id = local-1704369757182).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.5.0
      /_/

Using Scala version 2.12.18 (OpenJDK 64-Bit Server VM, Java 17.0.9)
Type in expressions to have them evaluated.
Type :help for more information.

scala> spark.time(new org.apache.spark.sql.execution.datasources.InMemoryFileIndex(spark, Seq(new org.apache.hadoop.fs.Path("s3a://dongjoon/100k/")), Map.empty, None))
24/01/04 04:02:45 WARN MetricsConfig: Cannot locate configuration: tried hadoop-metrics2-s3a-file-system.properties,hadoop-metrics2.properties
Time taken: 12310 ms
res0: org.apache.spark.sql.execution.datasources.InMemoryFileIndex = org.apache.spark.sql.execution.datasources.InMemoryFileIndex(s3a://dongjoon/100k)
```

**Apache Spark 4.0.0**
```
$ bin/spark-shell \
-c spark.hadoop.fs.s3a.aws.credentials.provider=com.amazonaws.auth.profile.ProfileCredentialsProvider
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 4.0.0-SNAPSHOT
      /_/

Using Scala version 2.13.12 (OpenJDK 64-Bit Server VM, Java 17.0.9)
Type in expressions to have them evaluated.
Type :help for more information.
24/01/04 04:06:26 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Spark context Web UI available at http://localhost:4040
Spark context available as 'sc' (master = local[*], app id = local-1704369986620).
Spark session available as 'spark'.

scala> spark.time(new org.apache.spark.sql.execution.datasources.InMemoryFileIndex(spark, Seq(new org.apache.hadoop.fs.Path("s3a://dongjoon/100k/")), Map.empty, None))
24/01/04 04:06:29 WARN MetricsConfig: Cannot locate configuration: tried hadoop-metrics2-s3a-file-system.properties,hadoop-metrics2.properties
24/01/04 04:06:29 WARN SDKV2Upgrade: Directly referencing AWS SDK V1 credential provider com.amazonaws.auth.profile.ProfileCredentialsProvider. AWS SDK V1 credential providers will be removed once S3A is upgraded to SDK V2
Time taken: 12857 ms
val res0: org.apache.spark.sql.execution.datasources.InMemoryFileIndex = org.apache.spark.sql.execution.datasources.InMemoryFileIndex(s3a://dongjoon/100k)
```

**Comment by cloud-fan** [comment]
@dongjoon-hyun  Thanks for the confirmation! I need to check with the benchmark team and verify if it's caused by extra storage layers (e.g. DBFS) not supporting this new API well.

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan !

**Pull request closed without merging**